### PR TITLE
Defensics: testing environment enhancement

### PR DESCRIPTION
**Pull request by KKopyscinski**

### CODE REVIEW

File: defensics/tooling.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import subprocess
+import threading
+import serial
+import os
+import time
+import logging
+import signal
+from queue import Queue
+
+pwd = ''
```
Comment (michal-narajowski) [inline defensics/tooling.py]: Overall the idea sounds good but lets use an env variable to pass the password to script.

Comment (michal-narajowski) [reply defensics/tooling.py]: Or use a configuration file.

File: defensics/tooling.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        """
+        Check if corefile exists; if board is not responding, restart it and check again
+        """
+        cmd = ['sudo', '-S', 'newtmgr', 'image', 'corelist', '-c', self.profile_name]
+        process = subprocess.Popen(cmd, shell=False, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+        output = process.communicate(input=pwd.encode())[0].decode()
+        logging.debug(output)
+        if 'Corefile present' in output:
+            logging.debug('Board has crashed, corefile present')
+            self.download_and_delete_corefile()
```
Comment (michal-narajowski) [inline defensics/tooling.py:151]: We are not handling errors here.

File: defensics/coap_automation_handler.py [hunk 3]
```diff
@@ -70,74 +59,129 @@ def process(self, job):
...
+                logging.debug(rc)
+                restart_try = 0
+                while rc == 1:
+                    logging.debug('restarting btmon')
+                    self.btmon.close()
+                    self.btmon = BTMonitor(testcase=test_case)
+
+                    rc = self.btmon.begin()
+                    logging.debug(rc)
+                    restart_try += 1
```
Comment (michal-narajowski) [inline defensics/coap_automation_handler.py:102]: Unused value

File: defensics/tooling.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        if self.testcase:
+            filename = str(self.testcase) + '.txt'
+        else:
+            filename = 'test.txt'
+        file = open(filename, 'a')
+        while not self.shutdown:
+            # limit data acquisition frequency
+            time.sleep(1)
+            bytesToRead = self.connection.inWaiting()
+            data = self.connection.read(bytesToRead).decode('latin-1')
```
[the remark this answers is not in the document]
Comment (KKopyscinski) [reply defensics/tooling.py:104]: Data is decoded to be logged in string format. Normally, a utf-8 would be used; however, some data after board boot remains in buffer. This data may be fragmented, causing errors like `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xea in position 0: invalid continuation byte` or `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1: invalid start byte`. This seemed to fix the issue, but as discussed offline, this buffered data will be consumed before run.

File: defensics/tooling.py (hunk 1 above)
Comment (KKopyscinski) [reply defensics/tooling.py]: As discussed offline, configuration file will be used

File: defensics/tooling.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        time.sleep(1)
+        bytesToRead = self.connection.inWaiting()
+        data = self.connection.readline(bytesToRead).decode()
+        logging.debug('Saving console output file')
+        file.write(data)
+        file.close()
+
+
+class NewtMgr:
+    def __init__(self, profile_name='test', type='oic_ble', connstring='peer_name=c5', testcase=None):
```
[the remark this answers is not in the document]
Comment (KKopyscinski) [reply defensics/tooling.py:120]: Agreed; now, they're mostly removed - only `testcase` defaults to `None`, and rest just requires `str` type.

File: defensics/tooling.py (hunk 2 above)
Comment (KKopyscinski) [reply defensics/tooling.py:151]: As errors here are not very dangerous, they're only being logged now

**Review by michal-narajowski — APPROVED** [review]

**Review by michal-narajowski — APPROVED** [review]

**Pull request merged**